syntax/std_seq_array_helper.go

```diff
@@ -0,0 +1,206 @@
+		} else {
+			finalVals = append(finalVals, a.Values()[absoluteIndex+1:]...)
+			break
+		}
+	}
+
+	return rel.NewArray(finalVals...)
+}
+
+// ArraySplit split a by b.
+func ArraySplit(a rel.Array, b rel.Value) rel.Value {
+	delimiter := convert2Array(b)
+	var result []rel.Value
+
+	if delimiter.Count() == 0 {
+		for _, e := range a.Values() {
+			result = append(result, rel.NewArray(e))
+		}
+	} else {
+		for start, absoluteIndex := 0, 0; start < a.Count(); {
+			relativeIndex := indexSubArray(a.Values()[start:], delimiter.Values())
+			if relativeIndex >= 0 {
+				absoluteIndex = relativeIndex + start
+				if start != absoluteIndex {
+					result = append(result, rel.NewArray(a.Values()[start:absoluteIndex]...))
+				}
+				start = absoluteIndex + delimiter.Count()
+			} else {
+				if start == 0 || start < a.Count() {
+					result = append(result, rel.NewArray(a.Values()[start:]...))
+				}
+				break
+			}
+		}
+	}
+
+	return rel.NewArray(result...)
+}
+
+// ArrayJoin joins array b to a, b is joiner and a is joinee.
+func ArrayJoin(a rel.Array, b rel.Value) rel.Value {
+	joiner := convert2Array(b)
+	if joiner.Count() == 0 || a.Count() == 0 {
+		// if joinee is empty, the final value will be empty
+		return a
+	}
+
+	vals := make([]rel.Value, 0, a.Count())
+	for i, value := range a.Values() {
+		switch vArray := value.(type) {
+		case rel.Array:
+			vals = append(vals, generate1LevelArray(vArray)...)
+		case rel.Value:
+			vals = append(vals, value)
+		}
+
+		if i+1 < a.Count() {
+			vals = append(vals, generate1LevelArray(joiner)...)
+		}
+	}
+
+	return rel.NewArray(vals...)
+}
+
+// ArrayPrefix check if a starts with b.
+func ArrayPrefix(a rel.Array, b rel.Value) rel.Value {
+	bArray := convert2Array(b)
+
+	if bArray.Count() == 0 {
+		return rel.NewBool(false)
+	}
+	if a.Count() < bArray.Count() {
+		return rel.NewBool(false)
+	}
+
+	bVals := bArray.Values()
+	bOffset := 0
+	arrayEnum, _ := a.ArrayEnumerator()
+	for arrayEnum.MoveNext() {
+		if bOffset < bArray.Count() && arrayEnum.Current().Equal(bVals[bOffset]) {
+			bOffset++
+			if bOffset == bArray.Count() {
+				break
+			}
+		} else {
+			return rel.NewBool(false)
+		}
+	}
+
+	return rel.NewBool(true)
+}
+
+// ArraySuffix check if a starts with b.
+func ArraySuffix(a rel.Array, b rel.Value) rel.Value {
+	bArray := convert2Array(b)
+
+	if bArray.Count() == 0 {
+		return rel.NewBool(false)
+	}
+	if a.Count() < bArray.Count() {
+		return rel.NewBool(false)
+	}
+
+	aVals := a.Values()
+	bVals := bArray.Values()
+	bOffset := bArray.Count() - 1
+
+	for _, val := range aVals[a.Count()-1:] {
+		if bOffset > -1 && val.Equal(bVals[bOffset]) {
+			bOffset--
+			if bOffset == -1 {
+				break
+			}
+		} else {
+			return rel.NewBool(false)
+		}
+	}
+
+	return rel.NewBool(true)
+}
+
+func convert2Array(val rel.Value) rel.Array {
+	switch val := val.(type) {
+	case rel.Array:
+		return val
+	case rel.GenericSet:
+		valArray, _ := rel.AsArray(val)
+		return valArray
+	case rel.Value:
```

Comment:
This method can try to convert rel.Array, rel.GenericSet and rel.Value to rel.Array only, otherwise it call panic. Actually this is a tool method to facilitate rel.Array conversion.

Reply:
Everything is a `rel.Value`, so the code below will simply wrap any value at all in an array and will never fail. Auto-converting this way is too clever and will result in confusing bugs.